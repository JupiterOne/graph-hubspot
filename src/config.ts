import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  oauthAccessToken: {
    type: 'string',
    mask: true,
  },
  apiBaseUrl: {
    type: 'string',
    mask: true,
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * This access_token is considered valid as a Bearer Authorization header
   */
  oauthAccessToken: string;

  /**
   * Hubspot API base url
   */
  apiBaseUrl: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;
  if (!Object.keys(instanceConfigFields).every((key) => config[key])) {
    throw new IntegrationValidationError(
      `Config requires all of {${Object.keys(instanceConfigFields).join(
        ', ',
      )}}`,
    );
  }

  const apiClient = createAPIClient(config, context.executionHistory);
  await apiClient.verifyAuthentication();
}
