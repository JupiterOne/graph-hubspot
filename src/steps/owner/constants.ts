export const OwnerEntity = {
  RESOURCE_NAME: 'Owner',
  TYPE: 'acme_user',
  CLASS: 'User',
};
