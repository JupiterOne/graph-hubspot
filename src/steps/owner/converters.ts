import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Owner } from '../../types';

export function createOwnerEntity(owner: Owner) {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: owner.id,
      },
      assign: {
        _key: `hubspot_owner:${owner.id}`,
        _type: 'acme_user',
        _class: 'User',
        // This is a custom property that is not a part of the data model class
        // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/User.json
        firstName: owner.firstName,
        // lastName: owner.lastName,
        name: `${owner.firstName} ${owner.lastName}`,
        email: owner.email,
        username: owner.email,
      },
    },
  });
}
