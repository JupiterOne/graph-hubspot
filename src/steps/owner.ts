import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';

export async function fetchOwners({
  jobState,
  instance,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);
  await apiClient.iterateOwners(async (owner) => {
    await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: owner,
          assign: {
            _key: 'acme-unique-account-id',
            _type: 'acme_account',
            _class: 'Account',
            mfaEnabled: true,
            // This is a custom property that is not a part of the data model class
            // hierarchy. See: https://github.com/JupiterOne/data-model/blob/master/src/schemas/Account.json
            manager: 'Manager Name',
          },
        },
      }),
    );
  });
}

export const ownerSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-owners',
    name: 'Fetch Owners',
    entities: [
      {
        resourceName: 'Owner',
        _type: 'acme_account',
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchOwners,
  },
];
