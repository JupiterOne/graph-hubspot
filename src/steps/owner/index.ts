import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../../client';
import { IntegrationConfig } from '../../config';
import { OwnerEntity } from './constants';
import { createOwnerEntity } from './converters';

export async function fetchOwners({
  jobState,
  instance,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);
  await apiClient.iterateOwners(async (owner) => {
    await jobState.addEntity(createOwnerEntity(owner));
  });
}

export const ownerSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-owners',
    name: 'Fetch Owners',
    entities: [
      {
        resourceName: OwnerEntity.RESOURCE_NAME,
        _type: OwnerEntity.TYPE,
        _class: OwnerEntity.CLASS,
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchOwners,
  },
];
