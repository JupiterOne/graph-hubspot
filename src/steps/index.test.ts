import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { integrationConfig } from '../../test/config';
import { IntegrationConfig } from '../config';
import { fetchOwners } from './owner';

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  // await fetchAccountDetails(context);
  await fetchOwners(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const owners = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(owners.length).toBeGreaterThan(0);
  expect(owners).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'acme_user' },
        firstName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['firstName'],
    },
  });
});
