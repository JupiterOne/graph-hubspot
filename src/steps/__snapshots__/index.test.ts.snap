// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "hubspot-owner-84186927",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "84186927",
          },
        },
      ],
      "_type": "acme_user",
      "createdOn": undefined,
      "displayName": "Stephen BOLIVAR",
      "email": "sb@tiringo.com",
      "firstName": "Stephen",
      "id": "84186927",
      "name": "Stephen BOLIVAR",
      "username": "sb@tiringo.com",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "acme_user",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
